src/v/raft/consensus.h

```diff
@@ -152,13 +152,17 @@ class consensus {
      * Last visible offset is updated in two scenarios
      *
      * - commited offset is updated (consistency_level=quorum)
-     * - when batch with was appended to the leader log with relaxed consistency
+     * - when batch that was appendend to the leader log is safely replicated on
+     *   majority of nodes
      *
      * We always update last visible index with std::max(prev,
      * possible_value) to guarantee its monotonicity.
      *
      */
-    model::offset last_visible_index() const { return _last_visible_index; };
+    model::offset last_visible_index() const {
+        return std::min(
+          _majority_replicated_index, _visibility_upper_bound_index);
```

Comment:
Why do we need `_visibility_upper_bound_index`? It seems we update it when we append to the disk so it's always ahead of `_majority_replicated_index` and this place is the only one when we read so `min` is redundant. I bet I missed something, @mmaslankaprv can you write please why `min` is necessary?

Reply:
So the `_visibility_upper_bound_index` serves as a 'switch' between relaxed and strong consistency modes that raft can operate in. When processing response from the follower (place where we update `_majority_replicated_index`) we do not have information about the consistency level that was requested by the `replicate` caller. Hence we need a way to differentiate if it was `quorum_ack` or relaxed mode. For quorum acks the upper bound is simply `committed_index` i.e. even tho majority of nodes replicate greater offset we can not read beyond that (it would cause linearizability violation). For relaxed consistency limit the `_visibility_upper_bound_index` is equal to log end offset.

Reply:
Thanks, got it!

Reply:
Wait there is an unexpected behaviour. Imagine that while the clients are trying to write `[a,b,c]` with `acks=-1` a leader gets isolated and an another client writes `[d]` with `acks=1`, it bumps `_visibility_upper_bound_index` and the never replicated messages `[a,b,c]` become visible.

Reply:
If we decide to this way we should reflect this behaviour in the documentation

Reply:
There are two indices that are taken into account when we deal with high watermark it is `_visibility_upper_bound_index` and either `commit_index` or `majority_replicated_offset`. In the example you mentioned the isolated leader will not increment neither  `commit_index` nor the `majority_replicated_offset`, hence the high watermark will not be updated.